### PR TITLE
Add Shared Role Distribution Chart

### LINKED ISSUE #2001: [TASK] Implement Validator's Role Distribution Pie Chart Section
## Description
Implement a pie chart to visualize the distribution of different roles within the Tangle Network's services on Validator Profile Page for each Validator. The pie chart should be interactive, allowing users to hover over segments to see additional details of each role.

Related tasks and dependency: https://github.com/webb-tools/webb-dapp/issues/1982
[Here](https://www.highcharts.com/demo/highcharts/variable-radius-pie) and [here](https://www.chartjs.org/docs/latest/charts/doughnut.html) are examples of potential hover effect with legend tooltip on the pie chart, for specific implementation please find an appropriate chart library. The color distribution of the pie chart is TBD and can be discussed with design @monaiuu 

### Screenshot
Image
Image

### PR DESCRIPTION
**Pull request by vutuanlinh2k2**
## Summary of changes

_Provide a detailed description of proposed changes._

- Add Shared Role Distribution Chart
- Refactor file names and getting data functions

### Proposed area of change

_Put an `x` in the boxes that apply._

- [ ] `apps/bridge-dapp`
- [ ] `apps/hubble-stats`
- [ ] `apps/stats-dapp`
- [x] `apps/tangle-dapp`
- [ ] `apps/testnet-leaderboard`
- [ ] `apps/faucet`
- [ ] `apps/zk-explorer`
- [ ] `libs/webb-ui-components`

### Reference issue to close (if applicable)

_Specify any issues that can be closed from these changes (e.g. `Closes #233`)._

- Closes #2001

### Screen Recording

https://github.com/webb-tools/webb-dapp/assets/69841784/8829b91b-ef22-40b5-ac11-86883bd1991e

CleanShot 2024-03-05 at 16 53 30@2x

---

### Code Checklist

_Please be sure to add .stories documentation if any additions are made to `libs/webb-ui-components`._

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: apps/tangle-dapp/components/charts/IndependentRoleDistributionChart.tsx
Comment (yurixander) [inline apps/tangle-dapp/components/charts/IndependentRoleDistributionChart.tsx:61]: If this is an amount derived from bigint (BN), eventually will need to be working with BN instead of numbers. Numbers can't hold all BN values.

File: apps/tangle-dapp/types/index.ts
Comment (yurixander) [inline apps/tangle-dapp/types/index.ts:148]: I've already implemented this on my PR, but should be fine. Once the merge occurs, should be resolved.

**Review by yurixander — APPROVED** [review]
Everything looks good. Left notes only, no changes. Some other things to consider:

1. We need consistent styling on the charts because I am making the same/similar charts for the restaking profile modals, and as of now there's minor differences in the styling.

2. For the data fed into the charts, or used anywhere, if they are token amounts or values, they should always be in BN form, otherwise it will crash when it tries to convert the amounts/balances received from Substrate into Numbers.

3. Because the amounts on the chart's data must be BNs, we have to get creative when feeding the data, since Recharts doesn't support BNs. What I did was instead of feeding it the amounts directly, I fed it proportions (0-1 float) representing percentages of the total. So example, if I have 100 TNT tokens in total, 65 of which are allocated for X role, then I feed `{X: Number(0.65)}` into Recharts. If you find another way of handing this, please let me know!

**Comment by vutuanlinh2k2** [comment]
> Everything looks good. Left notes only, no changes. Some other things to consider:
> 
> 1. We need consistent styling on the charts because I am making the same/similar charts for the restaking profile modals, and as of now there's minor differences in the styling.
> 2. For the data fed into the charts, or used anywhere, if they are token amounts or values, they should always be in BN form, otherwise it will crash when it tries to convert the amounts/balances received from Substrate into Numbers.
> 3. Because the amounts on the chart's data must be BNs, we have to get creative when feeding the data, since Recharts doesn't support BNs. What I did was instead of feeding it the amounts directly, I fed it percentages (0-1 float) of the total. So example, if I have 100 TNT tokens in total, 65 of which are allocated for X role, then I feed `{X: Number(0.65)}` into Recharts, and the remaining percentage can be calculated as `1 - 0.65`. If you find another way of handling this, please let me know!

Thanks for the comments! I will take a look at this. Also the CI is failing due to this [PR](https://github.com/webb-tools/webb-dapp/pull/2075) being merged, I will also investigate on this too

**Comment by vutuanlinh2k2** [comment]
> > Everything looks good. Left notes only, no changes. Some other things to consider:
> > 
> > 1. We need consistent styling on the charts because I am making the same/similar charts for the restaking profile modals, and as of now there's minor differences in the styling.
> > 2. For the data fed into the charts, or used anywhere, if they are token amounts or values, they should always be in BN form, otherwise it will crash when it tries to convert the amounts/balances received from Substrate into Numbers.
> > 3. Because the amounts on the chart's data must be BNs, we have to get creative when feeding the data, since Recharts doesn't support BNs. What I did was instead of feeding it the amounts directly, I fed it percentages (0-1 float) of the total. So example, if I have 100 TNT tokens in total, 65 of which are allocated for X role, then I feed `{X: Number(0.65)}` into Recharts, and the remaining percentage can be calculated as `1 - 0.65`. If you find another way of handling this, please let me know!
> 
> Thanks for the comments! I will take a look at this. Also the CI is failing due to this [PR](https://github.com/webb-tools/webb-dapp/pull/2075) being merged, I will also investigate on this too

@yurixander I've done some research and it seems there is no way to work with `BN` on `recharts`. I'm not sure why we have to calculate the percentage here. A workaround might be to convert the `BN` to `number` using `.toNumber()` method. This is not ideal with numbers between ﻿**-(2^53 - 1)** and ﻿**(2^53 - 1)** but the number outside of this range is quite extreme and most likely we're not gonna face it. Let me know what you think

**Comment by yurixander** [comment]
> > > Everything looks good. Left notes only, no changes. Some other things to consider:
> > > 
> > > 1. We need consistent styling on the charts because I am making the same/similar charts for the restaking profile modals, and as of now there's minor differences in the styling.
> > > 2. For the data fed into the charts, or used anywhere, if they are token amounts or values, they should always be in BN form, otherwise it will crash when it tries to convert the amounts/balances received from Substrate into Numbers.
> > > 3. Because the amounts on the chart's data must be BNs, we have to get creative when feeding the data, since Recharts doesn't support BNs. What I did was instead of feeding it the amounts directly, I fed it percentages (0-1 float) of the total. So example, if I have 100 TNT tokens in total, 65 of which are allocated for X role, then I feed `{X: Number(0.65)}` into Recharts, and the remaining percentage can be calculated as `1 - 0.65`. If you find another way of handling this, please let me know!
> > 
> > 
> > Thanks for the comments! I will take a look at this. Also the CI is failing due to this [PR](https://github.com/webb-tools/webb-dapp/pull/2075) being merged, I will also investigate on this too
> 
> @yurixander I've done some research and it seems there is no way to work with `BN` on `recharts`. I'm not sure why we have to calculate the percentage here. A workaround might be to convert the `BN` to `number` using `.toNumber()` method. This is not ideal with numbers between ﻿`-(2^53 - 1)` and ﻿`(2^53 - 1)` but the number outside of this range is quite extreme and most likely we're not gonna face it. Let me know what you think!

We should consider our specific case which is 18 decimals for the chain units. Larger amounts or more precise amounts than 16 decimals provided by Substrate (a totally valid and possible scenario) will cause problems in JavaScript when we try to convert them to numbers, likely throw an error.

As to what I meant for percentages, what we can do instead of providing Recharts the BN values, we provide it percentages of a total. These percentages are always within 0-1, so they can be accepted by Recharts.

The idea would be to have a total BN (ex max restaking amount), then per allocation, convert it to a percentage of that total (0-1). Then, after feeding the percentages to Recharts, the chart areas will be proportional to the amounts.

I've accomplished this on my open PR, [feel free to take a look at the `getPercentageOfTotal` function for reference](https://github.com/webb-tools/webb-dapp/pull/2053/files#diff-a493baf642652cee0a2c95cb08fcca9548d2756f6ed3318b063f971d948a7479).

image

**Comment by vutuanlinh2k2** [comment]
> > > > Everything looks good. Left notes only, no changes. Some other things to consider:
> > > > 
> > > > 1. We need consistent styling on the charts because I am making the same/similar charts for the restaking profile modals, and as of now there's minor differences in the styling.
> > > > 2. For the data fed into the charts, or used anywhere, if they are token amounts or values, they should always be in BN form, otherwise it will crash when it tries to convert the amounts/balances received from Substrate into Numbers.
> > > > 3. Because the amounts on the chart's data must be BNs, we have to get creative when feeding the data, since Recharts doesn't support BNs. What I did was instead of feeding it the amounts directly, I fed it percentages (0-1 float) of the total. So example, if I have 100 TNT tokens in total, 65 of which are allocated for X role, then I feed `{X: Number(0.65)}` into Recharts, and the remaining percentage can be calculated as `1 - 0.65`. If you find another way of handling this, please let me know!
> > > 
> > > 
> > > Thanks for the comments! I will take a look at this. Also the CI is failing due to this [PR](https://github.com/webb-tools/webb-dapp/pull/2075) being merged, I will also investigate on this too
> > 
> > 
> > @yurixander I've done some research and it seems there is no way to work with `BN` on `recharts`. I'm not sure why we have to calculate the percentage here. A workaround might be to convert the `BN` to `number` using `.toNumber()` method. This is not ideal with numbers between ﻿`-(2^53 - 1)` and ﻿`(2^53 - 1)` but the number outside of this range is quite extreme and most likely we're not gonna face it. Let me know what you think!
> 
> We should consider our specific case which is 18 decimals for the chain units. Larger amounts or more precise amounts than 16 decimals provided by Substrate (a totally valid and possible scenario) will cause problems in JavaScript when we try to convert them to numbers, likely throw an error.
> 
> As to what I meant for percentages, what we can do instead of providing Recharts the BN values, we provide it percentages of a total. These percentages are always within 0-1, so they can be accepted by Recharts.
> 
> The idea would be to have a total BN (ex max restaking amount), then per allocation, convert it to a percentage of that total (0-1). Then, after feeding the percentages to Recharts, the chart areas will be proportional to the amounts.
> 
> I've accomplished this on my open PR, [feel free to take a look at the `getPercentageOfTotal` function for reference](https://github.com/webb-tools/webb-dapp/pull/2053/files#diff-a493baf642652cee0a2c95cb08fcca9548d2756f6ed3318b063f971d948a7479).
> 
> image

Yes in this case we can merge this PR after @monaiuu review the UI. Thanks Yuri!

**Review by monaiuu — COMMENTED** [review]
Are we keeping this gradient color distribution to a separate PR? https://github.com/webb-tools/webb-dapp/issues/2001#issuecomment-1974240303

**Comment by vutuanlinh2k2** [comment]
> Are we keeping this gradient color distribution to a separate PR? [#2001 (comment)](https://github.com/webb-tools/webb-dapp/issues/2001#issuecomment-1974240303)

@monaiuu  I believe @AtelyPham is working on getting the service types with his current work. It would be the best if he implement it in his work

**Comment by yurixander** [comment]
> > Are we keeping this gradient color distribution to a separate PR? [#2001 (comment)](https://github.com/webb-tools/webb-dapp/issues/2001#issuecomment-1974240303)
> 
> @monaiuu I believe @AtelyPham is working on getting the service types with his current work. It would be the best if he implement it in his work

Note that I've added all the service types already on my open PR.

**Pull request merged**
